In [2]:
!pip list

Package              Version
-------------------- -----------
anyio                3.5.0
argon2-cffi          20.1.0
attrs                22.1.0
Babel                2.11.0
backcall             0.2.0
beautifulsoup4       4.11.1
bleach               4.1.0
brotlipy             0.7.0
certifi              2022.12.7
cffi                 1.15.1
charset-normalizer   2.0.4
colorama             0.4.6
cryptography         39.0.1
debugpy              1.5.1
decorator            5.1.1
defusedxml           0.7.1
entrypoints          0.4
fastjsonschema       2.16.2
idna                 3.4
importlib-metadata   4.11.3
importlib-resources  5.2.0
ipykernel            6.15.2
ipython              7.31.1
ipython-genutils     0.2.0
ipywidgets           7.6.5
jedi                 0.18.1
Jinja2               3.1.2
json5                0.9.6
jsonschema           4.17.3
jupyter              1.0.0
jupyter_client       7.4.9
jupyter-console      6.4.4
jupyter_core         4.11.2
jupyter-server       1.23.4
jupyte

In [3]:
!pip install torch torchvision torchaudio

  Using cached torchaudio-0.13.1-cp37-cp37m-win_amd64.whl (2.0 MB)


In [4]:
!pip install fastai

  Using cached fastai-2.7.12-py3-none-any.whl (233 kB)
  Using cached spacy-3.5.2-cp37-cp37m-win_amd64.whl (12.5 MB)
  Using cached scipy-1.7.3-cp37-cp37m-win_amd64.whl (34.1 MB)
  Using cached fastprogress-1.0.3-py3-none-any.whl (12 kB)
  Using cached matplotlib-3.5.3-cp37-cp37m-win_amd64.whl (7.2 MB)
  Using cached pandas-1.3.5-cp37-cp37m-win_amd64.whl (10.0 MB)
  Using cached scikit_learn-1.0.2-cp37-cp37m-win_amd64.whl (7.1 MB)
  Using cached PyYAML-6.0-cp37-cp37m-win_amd64.whl (153 kB)
  Using cached fastdownload-0.0.7-py3-none-any.whl (12 kB)
  Using cached fastcore-1.5.29-py3-none-any.whl (67 kB)
  Using cached pathy-0.10.1-py3-none-any.whl (48 kB)
  Using cached pydantic-1.10.7-cp37-cp37m-win_amd64.whl (2.1 MB)
  Using cached srsly-2.4.6-cp37-cp37m-win_amd64.whl (481 kB)
  Using cached catalogue-2.0.8-py3-none-any.whl (17 kB)
  Using cached cymem-2.0.7-cp37-cp37m-win_amd64.whl (30 kB)
  Using cached spacy_legacy-3.0.12-py2.py3-none-any.whl (29 kB)
  Using cached preshed-3.0.8-cp

## **PROGRAM BEGINS**

In [5]:
!pip install inltk

  Using cached inltk-0.9-py3-none-any.whl (13 kB)
  Using cached Bottleneck-1.3.7-cp37-cp37m-win_amd64.whl (109 kB)
  Using cached typing-3.7.4.3-py3-none-any.whl
  Using cached numexpr-2.8.4-cp37-cp37m-win_amd64.whl (92 kB)
  Using cached async_timeout-4.0.2-py3-none-any.whl (5.8 kB)
  Using cached nvidia_ml_py3-7.352.0-py3-none-any.whl
  Using cached aiohttp-3.8.4-cp37-cp37m-win_amd64.whl (322 kB)
  Using cached sentencepiece-0.1.98-cp37-cp37m-win_amd64.whl (977 kB)
  Using cached fastai-1.0.57-py3-none-any.whl (233 kB)
  Using cached asynctest-0.13.0-py3-none-any.whl (26 kB)
  Using cached aiosignal-1.3.1-py3-none-any.whl (7.6 kB)
  Using cached yarl-1.9.1-cp37-cp37m-win_amd64.whl (61 kB)
  Using cached frozenlist-1.3.3-cp37-cp37m-win_amd64.whl (34 kB)
  Using cached multidict-6.0.4-cp37-cp37m-win_amd64.whl (27 kB)
  Attempting uninstall: fastai
    Found existing installation: fastai 2.7.12
    Uninstalling fastai-2.7.12:
      Successfully uninstalled fastai-2.7.12


In [6]:
!pip install torch==1.3.1+cpu -f https://download.pytorch.org/whl/torch_stable.html

Looking in links: https://download.pytorch.org/whl/torch_stable.html
  Using cached https://download.pytorch.org/whl/cpu/torch-1.3.1%2Bcpu-cp37-cp37m-win_amd64.whl (71.3 MB)
  Attempting uninstall: torch
    Found existing installation: torch 1.13.1
    Uninstalling torch-1.13.1:
      Successfully uninstalled torch-1.13.1


ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
torchvision 0.14.1 requires torch==1.13.1, but you have torch 1.3.1+cpu which is incompatible.
torchaudio 0.13.1 requires torch==1.13.1, but you have torch 1.3.1+cpu which is incompatible.


### ***Run from here ->***

In [1]:
from inltk.inltk import setup

In [2]:
setup('hi')

RuntimeError: This event loop is already running

Done!


In [3]:
from inltk.inltk import tokenize

In [4]:
text = "इस पृष्ठ पर इन्टरनेट पर उपलब्ध विभिन्न हिन्दी एवं देवनागरी"
tokenize(text, 'hi')

['▁इस',
 '▁पृष्ठ',
 '▁पर',
 '▁इन्टरनेट',
 '▁पर',
 '▁उपलब्ध',
 '▁विभिन्न',
 '▁हिन्दी',
 '▁एवं',
 '▁देवनागरी']

## **INTERCEPTION POINT**

get_embedding_vectors() posed problem of:\
LSTM has no object attribute...\
RuntimeError: Internal: D:\a\sentencepiece\sentencepiece\src\sentencepiece_processor.cc(1102) [model_proto->ParseFromArray(serialized.data(), serialized.size())]

Resolved, how?\
*deactivated current env*\
*created new env*

Run always:\
!pip install torch torchvision torchaudio\
!pip install fastai\
!pip install inltk\
!pip install torch==1.3.1+cpu -f https://download.pytorch.org/whl/torch_stable.html\

In [5]:
from inltk.inltk import get_embedding_vectors

In [6]:
vectors = get_embedding_vectors(text, 'hi')
print(vectors)

[array([ 0.581913, -0.069027,  0.150849, -0.379967, ..., -0.259924, -0.144746,  0.325081,  0.100628], dtype=float32), array([-0.311319, -0.201469,  0.001034, -0.288219, ..., -0.250444,  0.135623, -0.108146, -0.284838], dtype=float32), array([-0.420775, -0.214894, -0.147413, -0.308402, ...,  0.594919,  0.344362,  0.505362, -0.434814], dtype=float32), array([ 0.331437, -0.197487, -0.262943, -0.706205, ...,  0.255144,  0.202683,  0.918716,  0.277362], dtype=float32), array([-0.420775, -0.214894, -0.147413, -0.308402, ...,  0.594919,  0.344362,  0.505362, -0.434814], dtype=float32), array([-0.12005 ,  0.035618, -0.295541, -0.298795, ..., -0.235442, -0.250848,  1.090632,  0.717407], dtype=float32), array([ 1.111676,  0.071516, -0.190193, -0.394667, ...,  0.072782, -0.280606,  0.532229,  0.680683], dtype=float32), array([ 0.25715 , -0.19085 ,  0.501046,  0.703017, ...,  0.132445,  0.167388,  0.525477, -0.04198 ], dtype=float32), array([-0.492057,  0.126205,  0.368837, -0.550006, ...,  1.4200

In [7]:
from inltk.inltk import predict_next_words

In [8]:
predict_next_words(text , 4, 'hi')

'इस पृष्ठ पर इन्टरनेट पर उपलब्ध विभिन्न हिन्दी एवं देवनागरी दोनो प्रकार के लिपियों'

In [24]:
# pip install --upgrade inltk

In [9]:
from inltk.inltk import get_sentence_encoding

In [10]:
get_sentence_encoding(text, 'hi')

array([-0.47964 , -0.058666, -0.014732,  0.988877, ..., -0.15301 , -0.014597, -0.700189, -0.027055], dtype=float32)

### ***Sentence similarity***

In [28]:
from inltk.inltk import get_sentence_similarity

s1 = "वे पेड़ के नीचे पत्ते खेल रहे हैं"
s2 = "वे पेड़ के नीचे ताश खेल रहे हैं"
s3 = "वे पेड़ के नीचे पर्ण खेल रहे हैं"

In [13]:
print(get_sentence_similarity(s1, s2, 'hi'))
print(get_sentence_similarity(s1, s3, 'hi'))
print(get_sentence_similarity(s2, s3, 'hi'))

0.9204173684120178


0.8794642090797424


0.8681066036224365


In [29]:
# s1 ans s2 are closest

### ***Similar sentences***

In [26]:
from inltk.inltk import get_similar_sentences

In [15]:
# s1 = "वे पेड़ के नीचे पत्ते खेल रहे हैं"
sim_sent = get_similar_sentences(s1, 10, 'hi', degree_of_aug = 0.1)
sim_sent

['वे पेड़ के परे पत्ते खेल रहे हैं',
 'वे पेड़ के बाहर पत्ते खेल रहे हैं',
 'वे पेड़ के आगे पत्ते खेल रहे हैं',
 'वे पेड़ संबंधी नीचे पत्ते खेल रहे हैं',
 'वे पेड़ के नीचे तने खेल रहे हैं',
 'वे पेड़ के अग्रभाग पत्ते खेल रहे हैं',
 'वे पेड़ के सम्मुख पत्ते खेल रहे हैं',
 'वह पेड़ के नीचे पत्ते खेल रहे हैं',
 'वे पेड़ के नीचे पत्तियां खेल रहे हैं',
 'वे पेड़ के नीचे पत्ते खेल रहा हैं']

In [39]:
# decreasing similarity

In [16]:
for sentence in sim_sent:
    print(get_sentence_similarity(s1, sentence, 'hi'))

0.9737790822982788


0.9692559838294983


0.9625504016876221


0.9565233588218689


0.9444694519042969


0.9432465434074402


0.9430104494094849


0.9403166770935059


0.928219199180603


0.9248740077018738


### ***Experiments***

In [41]:
# higher degree of augmentation => lesser similarity

In [17]:
# s1 = "वे पेड़ के नीचे पत्ते खेल रहे हैं"
sim_sent = get_similar_sentences(s1, 10, 'hi', degree_of_aug = 0.4)

for sentence in sim_sent:
    print(sentence, ":", get_sentence_similarity(s1, sentence, 'hi'))

वे पेड़ों के अग्रभाग पत्तियां खेल रहे हैं : 0.8788861036300659


उन वृक्षों में नीचे पत्ते खेल रहे हैं : 0.8481293320655823


वे वृक्ष के नीचे पत्ते खेला रहेहै : 0.8418086767196655


वे पेड़ से नीचे पत्तों क्रीड़ा रहे हैं : 0.8403995633125305


वे धुआं के सम्मुख पत्ते खेल चुके हैं : 0.8350328803062439


वे पेड़ के नीचे पत्तों क्रीड़ा थे हैं : 0.8313873410224915


वे चने सहित नीचे पत्ते खेल रहा हैं : 0.8084638714790344


वे पेड़ का परे पत्ते खेल कहलाए हैं : 0.8040140271186829


उन पेड़ में नीचे पत्ते खेल रहीं हैं : 0.7809008955955505


वे पेड़ को नीचे खंभे अपराधों रहे हैं : 0.7666933536529541


In [43]:
# s1 = "वे पेड़ के नीचे पत्ते खेल रहे हैं"
# s2 = "वे पेड़ के नीचे ताश खेल रहे हैं"
# s3 = "वे पेड़ के नीचे पर्ण खेल रहे हैं"

In [18]:
s1_list = get_similar_sentences(s1, 4, 'hi', degree_of_aug = 0.2)
s2_list = get_similar_sentences(s2, 4, 'hi', degree_of_aug = 0.2)
s3_list = get_similar_sentences(s3, 4, 'hi', degree_of_aug = 0.2)

In [19]:
s1_list

['वे पेड़ के उपर पत्ते खेल रहे हैं',
 'वे पेड़के नीचे पत्ते खेल रहे हैं',
 'वो पेड़ के नीचे पत्ते खेल रहे हैं',
 'वे पेड़ के नीचे पत्ते खेलों रहे हैं']

In [20]:
s2_list

['वे पेड़के नीचे ताश खेल रहे हैं',
 'वे पेड़ के नीचे ताश खेल रहे है',
 'वे पेड़ों के नीचे ताश खेल रहे हैं',
 'वो पेड़ के नीचे ताश खेल रहे हैं']

In [21]:
s3_list

['वे पेड़ के नीचे पर्ण खेल रहे है',
 'वे पेड़ के उपर पर्ण खेल रहे हैं',
 'वे पेड़ के नीचे पर्ण खेल रहें हैं',
 'वे पेड़ के नीचे पर्ण खेल रहें हैं']

In [49]:
# s1 = "वे पेड़ के नीचे पत्ते खेल रहे हैं"
# s2 = "वे पेड़ के नीचे ताश खेल रहे हैं"
# s3 = "वे पेड़ के नीचे पर्ण खेल रहे हैं"

##### Performing tests for similarity between s1_list and s2_list sentences

In [22]:
for sent1 in s1_list:
    print(sent1)
    for sent2 in s2_list:
        print("   -> ", sent2, ":", get_sentence_similarity(sent1, sent2, 'hi'))

वे पेड़ के उपर पत्ते खेल रहे हैं


   ->  वे पेड़के नीचे ताश खेल रहे हैं : 0.8825101256370544


   ->  वे पेड़ के नीचे ताश खेल रहे है : 0.9029024243354797


   ->  वे पेड़ों के नीचे ताश खेल रहे हैं : 0.8826947808265686


   ->  वो पेड़ के नीचे ताश खेल रहे हैं : 0.8581516742706299
वे पेड़के नीचे पत्ते खेल रहे हैं


   ->  वे पेड़के नीचे ताश खेल रहे हैं : 0.9206135272979736


   ->  वे पेड़ के नीचे ताश खेल रहे है : 0.8707882761955261


   ->  वे पेड़ों के नीचे ताश खेल रहे हैं : 0.8660560250282288


   ->  वो पेड़ के नीचे ताश खेल रहे हैं : 0.8680568337440491
वो पेड़ के नीचे पत्ते खेल रहे हैं


   ->  वे पेड़के नीचे ताश खेल रहे हैं : 0.8863694071769714


   ->  वे पेड़ के नीचे ताश खेल रहे है : 0.8695124387741089


   ->  वे पेड़ों के नीचे ताश खेल रहे हैं : 0.8630622625350952


   ->  वो पेड़ के नीचे ताश खेल रहे हैं : 0.9150738716125488
वे पेड़ के नीचे पत्ते खेलों रहे हैं


   ->  वे पेड़के नीचे ताश खेल रहे हैं : 0.8452292680740356


   ->  वे पेड़ के नीचे ताश खेल रहे है : 0.851597249507904


   ->  वे पेड़ों के नीचे ताश खेल रहे हैं : 0.8424384593963623


   ->  वो पेड़ के नीचे ताश खेल रहे हैं : 0.8089845180511475


In [23]:
# s1_list[1] and s2_list[0] yielded max. overlap

##### Example 2

In [36]:
sent1 = "इस किताब मे पन्ने की जानकारी दि है"
sent2 = "इस किताब मे हरित मण की जानकारी दि है"
sent3 = "इस किताब मे पना की जानकारी दि है"

In [37]:
sent1_list = get_similar_sentences(sent1, 4, 'hi', degree_of_aug = 0.2)
sent2_list = get_similar_sentences(sent2, 4, 'hi', degree_of_aug = 0.2)
sent3_list = get_similar_sentences(sent3, 4, 'hi', degree_of_aug = 0.2)

In [38]:
sent1_list

['इस गज़ल मे पन्ने की जानकारी दि है',
 'इस किताब मै पन्ने की जानकारी दि है',
 'इस किताब मे पन्ने का जानकारी दि है',
 'इस किताब मे मुखपृष्ठ की जानकारी दि है']

In [39]:
sent2_list

['इस वृत्तचित्र मे गंड मण की जानकारी दि है',
 'इस किताबमें हरित मण का जानकारी दि है',
 'इस किताब मे हरित विष्ट कि जानकारी दि है',
 'इस किताब मे मंदाकिनी ॥मण की जानकारी दि है']

In [40]:
sent3_list

['इस किताब मे पना का जानकारी दि है',
 'इस वृतांत मे पना की जानकारी दि है',
 'इसके किताब मे पना की जानकारी दि है',
 'इस किताब मे सिसना की जानकारी दि है']

In [41]:
for sent1 in sent1_list:
    print(sent1)
    for sent2 in sent2_list:
        print("   -> ", sent2, ":", get_sentence_similarity(sent1, sent2, 'hi'))

इस गज़ल मे पन्ने की जानकारी दि है


   ->  इस वृत्तचित्र मे गंड मण की जानकारी दि है : 0.8989073038101196


   ->  इस किताबमें हरित मण का जानकारी दि है : 0.8826530575752258


   ->  इस किताब मे हरित विष्ट कि जानकारी दि है : 0.830242931842804


   ->  इस किताब मे मंदाकिनी ॥मण की जानकारी दि है : 0.7541476488113403
इस किताब मै पन्ने की जानकारी दि है


   ->  इस वृत्तचित्र मे गंड मण की जानकारी दि है : 0.8685060143470764


   ->  इस किताबमें हरित मण का जानकारी दि है : 0.8710064888000488


   ->  इस किताब मे हरित विष्ट कि जानकारी दि है : 0.8188914656639099


   ->  इस किताब मे मंदाकिनी ॥मण की जानकारी दि है : 0.7433459758758545
इस किताब मे पन्ने का जानकारी दि है


   ->  इस वृत्तचित्र मे गंड मण की जानकारी दि है : 0.8507813811302185


   ->  इस किताबमें हरित मण का जानकारी दि है : 0.9088021516799927


   ->  इस किताब मे हरित विष्ट कि जानकारी दि है : 0.8057641983032227


   ->  इस किताब मे मंदाकिनी ॥मण की जानकारी दि है : 0.7384909987449646
इस किताब मे मुखपृष्ठ की जानकारी दि है


   ->  इस वृत्तचित्र मे गंड मण की जानकारी दि है : 0.9251208305358887


   ->  इस किताबमें हरित मण का जानकारी दि है : 0.8973046541213989


   ->  इस किताब मे हरित विष्ट कि जानकारी दि है : 0.8915088176727295


   ->  इस किताब मे मंदाकिनी ॥मण की जानकारी दि है : 0.7872669100761414


In [46]:
sent1 = "इस किताब मे पन्ने की जानकारी दि है"

print("   -> ", sent2_list[1], ":", get_sentence_similarity(sent2_list[1], sent1, 'hi'))
print("   -> ", sent2_list[0], ":", get_sentence_similarity(sent2_list[0], sent1, 'hi'))

   ->  इस किताबमें हरित मण का जानकारी दि है : 0.8825184106826782


   ->  इस वृत्तचित्र मे गंड मण की जानकारी दि है : 0.8925307989120483


In [47]:
s = "इस किताब में हरित मण का जानकारी दि है"

In [48]:
get_sentence_similarity(s, sent1, 'hi')

0.8678280115127563

#### Experimentation

### ***Lemmatization***

In [50]:
!pip install nltk

  Using cached nltk-3.8.1-py3-none-any.whl (1.5 MB)
     -------------------------------------- 268.0/268.0 kB 1.0 MB/s eta 0:00:00


In [11]:
# -*- coding: utf-8 -*-
import io
import string
import re
import nltk
from nltk.tag import tnt
from nltk.corpus import indian


outputFile = open('testOutput.txt','w')
wordDict = {}
taggedFile  = open("output_test.txt", "r")

def checkRules():
    rules = []
    with open('rules.txt','r') as rulesFile:
        for line in rulesFile:
            line = line.strip('\n').split()
            rules.append(line)

    rulesTwo = []
    with open('rules2.txt','r') as rulesFile:
        for line in rulesFile:
            line = line.strip('\n').split()
            rulesTwo.append(line)


    taggedData = []
    prev= ['None']
    for line in taggedFile:
        prevTag = prev.pop(0)
        line= line.strip('\n').split()
        line.append(prevTag)
        currTag = line[1]
        prev.append(currTag)
        taggedData.append(line)
   
    for element in taggedData:
        if element[0]=='<s>' or element[0]=='</s>':
            
            outputFile.write(element[0] + ' ' + element[1])
            outputFile.write('\n')
            continue
        if any(char.isdigit() for char in element[0]) or element[1] == 'SYM':

            outputFile.write(element[0] + ' ' + element[0])
            outputFile.write('\n')
            continue
        flag = 0
        count = 0
        match = []
        for rule in rules:
            if element[:2] == rule[:2]:
                match.append(rule[2])
                count = count +1 

        if count==1:
            outputFile.write(element[0] + ' ' + match[0])
            outputFile.write('\n') 
            flag = 1
        
        elif count>1:
            for rule in rulesTwo:
                if element == rule[:3]:
                    outputFile.write(element[0] + ' ' + rule[3])
                    outputFile.write('\n')
                    flag = 1
                    break

        if flag == 0:
            lemma = lemmatize(element)
            outputFile.write(element[0] + ' ' + lemma)
            outputFile.write('\n')
    
def lemmatize(element):

        if element[1].strip().startswith("NN") or element[1].strip().startswith("VAUX") :
            return element[0]

        else:
        
            return generate_stem_words(element[0].strip())
    

def generate_stem_words(word):
    
    suffixes = {
    1: ["ो","े","ू","ु","ी","ि","ा"],
    2: ["कर","ाओ","िए","ाई","ाए","ने","नी","ना","ते","ीं","ती","ता","ाँ","ां","ों","ें"],
    3: ["ाकर","ाइए","ाईं","ाया","ेगी","ेगा","ोगी","ोगे","ाने","ाना","ाते","ाती","ाता","तीं","ाओं","ाएं","ुओं","ुएं","ुआं"],
    4: ["ाएगी","ाएगा","ाओगी","ाओगे","एंगी","ेंगी","एंगे","ेंगे","ूंगी","ूंगा","ातीं","नाओं","नाएं","ताओं","ताएं","ियाँ","ियों","ियां"],
    5: ["ाएंगी","ाएंगे","ाऊंगी","ाऊंगा","ाइयाँ","ाइयों","ाइयां"],
        }


    rule1 = suffixes[4][17]
    rule2 = suffixes[4][16]
    rule3 = suffixes[1][1]
    rule4 = suffixes[1][6]
    rule5 = suffixes[1][4]
    rule6 = suffixes[5][4]
    rule7 = suffixes[5][5]
    rule8 = suffixes[5][6]

    
    if word.endswith(rule1):
        return word[:word.rindex(rule1)] + suffixes[1][4]
    
    elif word.endswith(rule2):
        return word[:word.rindex(rule2)] + suffixes[1][4]

    elif word.endswith(rule3):
        return word[:word.rindex(rule3)] + suffixes[1][6]
    
    elif word.endswith(rule4):
        return word[:word.rindex(rule4)] + suffixes[1][6]
    
    elif word.endswith(rule5):
        return word[:word.rindex(rule5)] + suffixes[1][4]
    
    elif word.endswith(rule6):
        return word[:word.rindex(rule6)] + suffixes[3][2]
    
    elif word.endswith(rule7):
        return word[:word.rindex(rule7)] + suffixes[3][2]
    
    elif word.endswith(rule8):
        return word[:word.rindex(rule8)] + suffixes[3][2]
    
    else:
        
        for key in suffixes.keys():
            for value in suffixes[key]:
                if word.endswith(value):
                    suff = word.rindex(value)
                    return word[:suff]

    return word
     
    

def main():
    checkRules()
     
        
if __name__=="__main__":
    main()

FileNotFoundError: [Errno 2] No such file or directory: 'output_test.txt'

### ***pyiwn***

In [54]:
!pip install --upgrade pyiwn

In [14]:
import pyiwn

In [15]:
iwn = pyiwn.IndoWordNet()

2023-04-26:09:43:42,802 INFO     [iwn.py:43] Loading hindi language synsets...


In [16]:
word = 'पत्ता'
synsets = iwn.synsets(word)

In [17]:
synsets

[Synset('पत्ता.noun.2535'),
 Synset('चादर.noun.5193'),
 Synset('ताश.noun.5989'),
 Synset('पीपरपर्न.noun.6170'),
 Synset('पत्ता.noun.20321'),
 Synset('पत्ता.noun.34975')]

In [18]:
for synset in synsets:
    print(synset.lemma_names()[0])

पत्ता
चादर
ताश
पीपरपर्न
पत्ता
पत्ता


##### Synset synonym overlap with max. yield sentences

In [34]:
# s1_list[1]: वे पेड़के नीचे पत्ते खेल रहे हैं
# s2_list[0]: वे पेड़के नीचे ताश खेल रहे हैं

for synset in synsets:
    w = synset.lemma_names()[0]
    print(synset.lemma_names())
    for wrd in synset.lemma_names():
        if wrd != word:
            w = wrd
            break
    print(w)
    print("s1:", get_sentence_similarity(s1_list[1], w, 'hi'))
    print("s2:", get_sentence_similarity(s2_list[0], w, 'hi'))
    print("--")

['पत्ता', 'पात', 'पर्ण', 'पत्र', 'दल', 'छद', 'पत्रक', 'परन', 'वर्ह']
पात


s1: 0.07805177569389343


s2: 0.11630647629499435
--
['चादर', 'पत्तर', 'फलक', 'चद्दर', 'पत्ता', 'पत्तरा', 'धातु-पटल', 'धातु-पत्र', 'धातुपटल', 'धातु पटल', 'धातु पत्र', 'धातुपत्र', 'पत्तरी', 'पत्ती']
चादर


s1: 0.06818144768476486


s2: 0.08236005902290344
--
['ताश', 'पत्ता', 'कार्ड', 'तास', 'गंजीफा', 'गंजीफ़ा']
ताश


s1: -0.031271252781152725


s2: -0.015416072681546211
--
['पीपरपर्न', 'पत्ता', 'झुलमुली', 'झाला', 'पीपलपत्ता']
पीपरपर्न


s1: -0.08269177377223969


s2: -0.09443926066160202
--
['पत्ता']
पत्ता


s1: 0.1224944144487381


s2: 0.10395534336566925
--
['पत्ता']
पत्ता


s1: 0.1224944144487381


s2: 0.10395534336566925
--


In [ ]:
##### Synset synonym overlap with max. yield sentences

In [79]:
# s1_list[1]: वे पेड़के नीचे पत्ते खेल रहे हैं
# s2_list[0]: वे पेड़के नीचे ताश खेल रहे हैं

for synset in synsets:
    print(synset.lemma_names())
    for sent in synset.examples():
        print("s1:", get_sentence_similarity(s1, sent, 'hi'))
        print("s2:", get_sentence_similarity(s2, sent, 'hi'))
    print("--")

['पत्ता', 'पात', 'पर्ण', 'पत्र', 'दल', 'छद', 'पत्रक', 'परन', 'वर्ह']


s1: 0.20420344173908234


s2: 0.1643276810646057
--
['चादर', 'पत्तर', 'फलक', 'चद्दर', 'पत्ता', 'पत्तरा', 'धातु-पटल', 'धातु-पत्र', 'धातुपटल', 'धातु पटल', 'धातु पत्र', 'धातुपत्र', 'पत्तरी', 'पत्ती']


s1: 0.09580745548009872


s2: 0.08708316832780838
--
['ताश', 'पत्ता', 'कार्ड', 'तास', 'गंजीफा', 'गंजीफ़ा']


s1: 0.06619558483362198


s2: 0.047604985535144806


s1: 0.1149706095457077


s2: 0.12649911642074585
--
['पीपरपर्न', 'पत्ता', 'झुलमुली', 'झाला', 'पीपलपत्ता']


s1: 0.15046444535255432


s2: 0.11927741765975952
--
['पत्ता']


s1: 0.1286919265985489


s2: 0.09179813414812088


s1: 0.18547263741493225


s2: 0.16454263031482697
--
['पत्ता']


s1: 0.13931410014629364


s2: 0.10987623035907745
--


In [52]:
s1_list

['वे पेड़ के उपर पत्ते खेल रहे हैं',
 'वे पेड़के नीचे पत्ते खेल रहे हैं',
 'वो पेड़ के नीचे पत्ते खेल रहे हैं',
 'वे पेड़ के नीचे पत्ते खेलों रहे हैं']

In [19]:
word = 

SyntaxError: invalid syntax (2336514174.py, line 1)

In [48]:
word = 'पत्ता'
sentence = "वे पेड़ के नीचे पत्ते खेल रहे हैं"
synsets = iwn.synsets(word)
s = []

for synset in synsets:
    w = synset.lemma_names()[0]
    # print(synset.lemma_names())
    for wrd in synset.lemma_names():
        if wrd != word:
            s.append(sentence.replace("पत्ते", wrd))
            break
            
s

['वे पेड़ के नीचे पात खेल रहे हैं',
 'वे पेड़ के नीचे चादर खेल रहे हैं',
 'वे पेड़ के नीचे ताश खेल रहे हैं',
 'वे पेड़ के नीचे पीपरपर्न खेल रहे हैं']

In [24]:
# get similar sentences of these four sentences

In [52]:
for sent in s:
    sim_sent_list = get_similar_sentences(sent, 4, 'hi', degree_of_aug = 0.2)
    for sen in sim_sent_list:
        print(sen, ":", get_sentence_similarity(sen, sentence, 'hi'))
    print()

वे पेड़ के नीचे पात खेल रहे है : 0.8769082427024841


वे पेड़ों के नीचे पात खेल रहे हैं : 0.8740048408508301


वे पेड़ के नीचे झूला खेल रहे हैं : 0.8995546698570251


वे पेड़ के नीचे शिखा खेल रहे हैं : 0.9245168566703796



वे पेड़ के नीचे चादर खेल रहे है : 0.8997978568077087


वे पेड़ की नीचे चादर खेल रहे हैं : 0.9247074127197266


वे पेड़ के उपर चादर खेल रहे हैं : 0.900107741355896


वे पेड़ के नीचे चादर खेल रहें हैं : 0.8926532864570618



वे पेड़के नीचे ताश खेल रहे हैं : 0.9055671095848083


वे पेड़के नीचे ताश खेल रहे हैं : 0.9055671095848083


वे पेड़ की नीचे ताश खेल रहे हैं : 0.9218146204948425


वो पेड़ के नीचे ताश खेल रहे हैं : 0.8833142518997192



वे पेड़के ऊपर पीपरपर्न खेल रहे हैं : 0.8199869990348816


वे पेड़ के उपर पी्सेपर्न खेल रहे हैं : 0.8609452843666077


वो पेड़ के नीचे पीपरपसर्न खेल रहे हैं : 0.7994949817657471


वे पेड़ के नीचे पीपरप्रर्न खेल रहे है : 0.8322130441665649



##### Results

वे पेड़ के नीचे पात खेल रहे है : 0.8769082427024841\
वे पेड़ों के नीचे पात खेल रहे हैं : 0.8740048408508301\
वे पेड़ के नीचे झूला खेल रहे हैं : 0.8995546698570251\
**वे पेड़ के नीचे शिखा खेल रहे हैं : 0.9245168566703796**\
\
वे पेड़ के नीचे चादर खेल रहे है : 0.8997978568077087\
**वे पेड़ की नीचे चादर खेल रहे हैं : 0.9247074127197266**\
**वे पेड़ के उपर चादर खेल रहे हैं : 0.900107741355896**\
वे पेड़ के नीचे चादर खेल रहें हैं : 0.8926532864570618\
\
**वे पेड़के नीचे ताश खेल रहे हैं : 0.9055671095848083**\
***वे पेड़के नीचे ताश खेल रहे हैं : 0.9055671095848083***\
**वे पेड़ की नीचे ताश खेल रहे हैं : 0.9218146204948425**\
वो पेड़ के नीचे ताश खेल रहे हैं : 0.8833142518997192\
\
वे पेड़के ऊपर पीपरपर्न खेल रहे हैं : 0.8199869990348816\
वे पेड़ के उपर पी्सेपर्न खेल रहे हैं : 0.8609452843666077\
वो पेड़ के नीचे पीपरपसर्न खेल रहे हैं : 0.7994949817657471\
वे पेड़ के नीचे पीपरप्रर्न खेल रहे है : 0.8322130441665649